src/test/java/BaseTest.java

```diff
@@ -1,11 +1,166 @@
 import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
 import org.openqa.selenium.chrome.ChromeOptions;
-import org.testng.annotations.BeforeSuite;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.Assert;
+import org.testng.annotations.*;
+
+import java.util.*;
+
+import java.time.Duration;
 
 public class BaseTest {
 
+    public static WebDriver driver = null;
+    public static WebDriverWait wait = null;
+    public static Actions actions = null;
+    public static String url = "https://qa.koel.app/";
+
     @BeforeSuite
     static void setupClass() {
         WebDriverManager.chromedriver().setup();
     }
+    @BeforeMethod
+    @Parameters ({"BaseURL"})
+    public void launchBrowser(String BaseURL) throws InterruptedException{
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("--remote-allow-origins=*");
+
+        driver = new ChromeDriver(options);
+        url = BaseURL;
+        driver.get(url);
+        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
+        actions = new Actions(driver);
+
+    }
+    @AfterMethod
+    public void logout() {
+
+        driver.quit();
+    }
+    @DataProvider (name="CorrectLoginProviders")
+    public static Object[][] getDataFromDataProviders(){
+        return new Object[][] {
+                {"matt.pierce@testpro.io", "te$t$tudent"}
+        };
+    }
+
+    public static void openLoginUrl() {
+        String url = "https://qa.koel.app/";
+        driver.get(url);
+    }
+
+    protected static void enterEmail(String email) throws InterruptedException{
+        WebElement emailInput = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[type='email']")));
+        emailInput.click();
+        emailInput.clear();
+        //inputs valid email address into email input field
+        emailInput.sendKeys(email);
+    }
+
+    protected static void enterPassword(String password) throws InterruptedException{
+
+        WebElement passwordInput = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[type='password']")));
+        passwordInput.clear();
+        //inputs valid password for login into password field
+        passwordInput.sendKeys(password);
+    }
+
+    public void clickSubmit() throws InterruptedException{
+        WebElement loginButton = wait.until(ExpectedConditions.elementToBeClickable((By.cssSelector("[type='submit']"))));
+        loginButton.click();
+    }
```

Comment:
Page Object Model (POM): recommended to associate the methods with the corresponding page class where they are relevant. In the POM, each page or section of a web application is represented by a separate class. These classes encapsulate the elements and actions specific to that page or section.

Move these methods to the corresponding class according to the POM.